src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -64,204 +64,213 @@ public KestrelMetrics(IMeterFactory meterFactory)
+    public void ConnectionStop(in ConnectionMetricsContext metricsContext, Exception? exception, List<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
     {
-        if (_currentConnectionsCounter.Enabled || _connectionDuration.Enabled)
+        if (metricsContext.CurrentConnectionsCounterEnabled || metricsContext.ConnectionDurationEnabled)
         {
-            ConnectionStopCore(connection, exception, customTags, startTimestamp, currentTimestamp);
+            ConnectionStopCore(metricsContext, exception, customTags, startTimestamp, currentTimestamp);
         }
     }
 
     [MethodImpl(MethodImplOptions.NoInlining)]
-    private void ConnectionStopCore(BaseConnectionContext connection, Exception? exception, List<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    private void ConnectionStopCore(in ConnectionMetricsContext metricsContext, Exception? exception, List<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
     {
         var tags = new TagList();
-        InitializeConnectionTags(ref tags, connection);
+        InitializeConnectionTags(ref tags, metricsContext);
 
-        // Decrease in connections counter must match tags from increase. No custom tags.
-        _currentConnectionsCounter.Add(-1, tags);
-
-        if (exception != null)
+        if (metricsContext.CurrentConnectionsCounterEnabled)
         {
-            tags.Add("exception-name", exception.GetType().FullName);
+            // Decrease in connections counter must match tags from increase. No custom tags.
+            _currentConnectionsCounter.Add(-1, tags);
         }
 
-        // Add custom tags for duration.
-        if (customTags != null)
+        if (metricsContext.ConnectionDurationEnabled)
         {
-            for (var i = 0; i < customTags.Count; i++)
+            if (exception != null)
             {
-                tags.Add(customTags[i]);
+                tags.Add("exception-name", exception.GetType().FullName);
             }
-        }
 
-        var duration = Stopwatch.GetElapsedTime(startTimestamp, currentTimestamp);
-        _connectionDuration.Record(duration.TotalSeconds, tags);
+            // Add custom tags for duration.
+            if (customTags != null)
+            {
+                for (var i = 0; i < customTags.Count; i++)
+                {
+                    tags.Add(customTags[i]);
+                }
+            }
+
+            var duration = Stopwatch.GetElapsedTime(startTimestamp, currentTimestamp);
+            _connectionDuration.Record(duration.TotalSeconds, tags);
+        }
     }
 
-    public void ConnectionRejected(BaseConnectionContext connection)
+    public void ConnectionRejected(in ConnectionMetricsContext metricsContext)
     {
         if (_rejectedConnectionsCounter.Enabled)
```

Comment:
```suggestion
        // Check live rather than cached state because this is just a counter, it's not a start/stop event like the other metrics.
        if (_rejectedConnectionsCounter.Enabled)
```